### ISSUE TITLE
formatlist is not working

### ISSUE THREAD

**Issue by jwaldrip** [hashicorp/terraform#7607]
I have the following code, and I am getting an error regarding formatlist. Why wouldn't this be working?

**Terraform: v0.6.16**
**OS: Mac**

**Code**:

``` hcl
resource "template_file" "etcd-write_files" {
  template = "${file("config/write_files/etcd.yml")}"
}

resource "template_file" "etcd-units" {
  template = "${file("config/units/etcd.yml")}"
}

module "etcd-coreos-user-data" {
  depends_on                        = ["google_compute_address.etcd.*"]
  source                            = "git::https://github.com/brandfolder/terraform-coreos-user-data.git?ref=master"
  etcd2_initial-cluster             = "${join(",", formatlist("%s=https://%s:%s/", google_compute_address.etcd.*.name, google_compute_address.etcd.*.address, "2380"))}"
  etcd2_advertise-client-urls       = "http://var!private_ipv4:2379,http://var!private_ipv4:4001"
  etcd2_initial-advertise-peer-urls = "http://var!private_ipv4:2380,http://var!private_ipv4:7001"
  etcd2_listen-client-urls          = "http://0.0.0.0:2379,http://0.0.0.0:4001"
  etcd2_listen-peer-urls            = "http://var!private_ipv4:2380,http://var!private_ipv4:7001"
  etcd2_debug                       = "true"
  flannel_interface                 = "var!private_ipv4"
  fleet_metadata                    = "role=etcd"
  fleet_public_ip                   = "var!private_ipv4"
  fleet_engine_reconcile_interval   = "10"
  fleet_etcd_request_timeout        = "5.0"
  fleet_agent_ttl                   = "120s"
  write_files                       = "${template_file.etcd-write_files.rendered}"
  units                             = "${template_file.etcd-units.rendered}"
}

resource "google_compute_disk" "etcd" {
  count = "${var.etcd-count}"
  name  = "${replace("${var.prefix}-etcd-${count.index}", "/^-/", "")}"
  zone  = "${element(split(",", var.zones), count.index % length(split(",", var.zones)))}"
  image = "${coalesce(var.etcd-image, var.default-image)}"
  type  = "pd-ssd"
  size  = 100

  lifecycle {
    # Ignore images changes so that we dont ever delete the disk when updating
    # to a later version of CoreOS
    ignore_changes = ["image"]
  }
}

resource "google_compute_address" "etcd" {
  count  = "${var.etcd-count}"
  name   = "${replace("${var.prefix}-etcd-${count.index}", "/^-/", "")}"
  region = "${replace(element(split(",", var.zones), count.index % length(split(",", var.zones))), "/-[a-z]$/", "")}"
}

resource "google_compute_instance" "etcd" {
  count        = "${var.etcd-count}"
  name         = "${replace("${var.prefix}-etcd-${count.index}", "/^-/", "")}"
  description  = "Etcd master"
  machine_type = "${coalesce(var.etcd-instance-type, var.default-instance-type)}"
  zone         = "${element(split(",", var.zones), count.index % length(split(",", var.zones)))}"

  tags = ["etcd"]

  disk {
    disk        = "${element(google_compute_disk.etcd.*.name, count.index)}"
    auto_delete = false
  }

  scheduling {
    automatic_restart   = true
    on_host_maintenance = "MIGRATE"
  }

  network_interface {
    subnetwork = "${element(google_compute_subnetwork.primary.*.name, count.index % length(split(",", var.zones)))}"

    access_config {
      nat_ip = "${element(google_compute_address.etcd.*.address, count.index)}"
    }
  }

  metadata {
    user-data = "${module.etcd-coreos-user-data.user-data}"
  }

  service_account {
    scopes = ["userinfo-email", "compute-ro", "storage-ro"]
  }
}
```

**Error:**

```
* formatlist: no lists in arguments to formatlist in:

${join(",", formatlist("%s=https://%s:%s/", google_compute_address.etcd.*.name, google_compute_address.etcd.*.address, "2380"))}
* formatlist: no lists in arguments to formatlist in:
```

**Comment by brikis98** [comment]
Just hit this error in Terraform 0.7.2. Here is a tiny example to repro:

```hcl
data "template_file" "foo" {
  template = "${join(", ", formatlist("%s", var.list))}"
}

variable "list" {
  type = "list"
  default = []
}

output "foo" {
  value = "${data.template_file.foo.rendered}"
}
```

When you run it, you get:

```
Errors:

  * formatlist: no lists in arguments to formatlist in:

${formatlist("%s", var.list)}
```

The cause is logic in [this function](https://github.com/hashicorp/terraform/blob/81bb6b72648135347803f4cdad529b08e43e7347/config/interpolate_funcs.go#L430) that counts the number of items in the list(s) passed to `formatlist`. The problem is that 0 items could mean "you didn't pass in any lists" (which is a legitimate error) or "your lists are empty" (which is perfectly valid).

**Comment by mitchellh** [comment]
Fix in #9795
